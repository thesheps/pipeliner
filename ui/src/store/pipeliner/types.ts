export const SHOW_ERROR = "SHOW_ERROR";
export const HIDE_ERROR = "HIDE_ERROR";

export interface ShowErrorAction {
  type: typeof SHOW_ERROR;
  errorMessage: string;
}

export interface HideErrorAction {
  type: typeof HIDE_ERROR;
}

export type PipelinerActionTypes = ShowErrorAction | HideErrorAction;

export interface PipelinerState {
  showError: boolean;
  errorMessage: string;
}
