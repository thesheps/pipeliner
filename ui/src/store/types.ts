import { UIState } from "./ui/types";
import { UserState } from "./user/types";

export interface PipelinerState {
  ui: UIState;
  user: UserState;
}
