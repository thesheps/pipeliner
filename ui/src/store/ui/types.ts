export const SET_SHOW_REGISTER_MODAL = "SET_SHOW_REGISTER_MODAL";
export const SHOW_ERROR = "SHOW_ERROR";
export const HIDE_ERROR = "HIDE_ERROR";

export interface ShowErrorAction {
  type: typeof SHOW_ERROR;
  errorMessage: string;
}

export interface HideErrorAction {
  type: typeof HIDE_ERROR;
}

export interface SetShowRegisterModalAction {
  type: typeof SET_SHOW_REGISTER_MODAL;
  show: boolean;
}

export type UIActionTypes =
  | ShowErrorAction
  | HideErrorAction
  | SetShowRegisterModalAction;

export interface UIState {
  showRegisterModal: boolean;
  showError: boolean;
  errorMessage: string;
}
