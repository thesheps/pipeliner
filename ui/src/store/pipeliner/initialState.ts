import { PipelinerState } from "./types";

export const initialPipelinerState: PipelinerState = {
  showError: false,
  errorMessage: "",
};
