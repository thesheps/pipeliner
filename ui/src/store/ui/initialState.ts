import { UIState } from "./types";

export const initialUIState: UIState = {
  showError: false,
  errorMessage: "",
};
