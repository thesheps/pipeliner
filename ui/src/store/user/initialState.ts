import { UserState } from "./types";

export const initialUserState: UserState = {
  authToken: "",
  isAuthenticating: false,
  isSignedIn: false,
};
