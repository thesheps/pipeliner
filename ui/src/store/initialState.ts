import { initialUserState } from "./user/initialState";
import { initialPipelinerState } from "./pipeliner/initialState";

export const initialState = {
  ...initialUserState,
  ...initialPipelinerState,
};
