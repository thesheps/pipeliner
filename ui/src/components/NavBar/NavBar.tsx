import React from "react";
import MenuIcon from "@material-ui/icons/Menu";
import { connect } from "react-redux";
import { makeStyles } from "@material-ui/core/styles";
import { AppBar, Toolbar, IconButton, Typography } from "@material-ui/core";

import { UserActions } from "../UserActions";
import { PipelinerState } from "../../store/types";
import { userSelector } from "../../store/user/selectors";

export interface NavBarProps {
  appName: string;
  isSignedIn: boolean;
}

interface StateProps {
  isSignedIn: boolean;
}

const useStyles = makeStyles((theme) => ({
  root: { flexGrow: 1 },
  menuButton: { marginRight: theme.spacing(2) },
  title: { flexGrow: 1 },
}));

export const NavBar = ({ appName, isSignedIn }: NavBarProps) => {
  const classes = useStyles();

  return (
    <AppBar position="static">
      <Toolbar>
        <IconButton
          edge="start"
          className={classes.menuButton}
          color="inherit"
          aria-label="menu"
        >
          <MenuIcon />
        </IconButton>

        <Typography variant="h6" className={classes.title}>
          {appName}
        </Typography>

        <UserActions data-testid="user-actions" isSignedIn={isSignedIn} />
      </Toolbar>
    </AppBar>
  );
};

const mapStateToProps = (state: PipelinerState): StateProps => {
  const userState = userSelector(state);

  return {
    isSignedIn: userState.isSignedIn,
  };
};

export const NavBarContainer = connect(mapStateToProps)(NavBar);
