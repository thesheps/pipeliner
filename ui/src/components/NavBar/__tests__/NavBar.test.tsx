import React from "react";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { fireEvent, render } from "@testing-library/react";

import {
  SET_IS_AUTHENTICATING,
  UserState,
  SIGN_IN,
} from "../../../store/user/types";
import { SET_SHOW_REGISTER_MODAL, SHOW_SUCCESS } from "../../../store/ui/types";
import { UserService } from "../../../services";
import { NavBar, NavBarContainer } from "../NavBar";
import { initialState } from "../../../store/initialState";

jest.mock("../../../services");

describe("NavBar", () => {
  const appName = "My Lovely Test";
  const username = "joe.bloggs";
  const emailAddress = "joe.bloggs@invalid.com";
  const password = "my super awesome password";
  const mockStore = configureStore([thunk]);

  it("renders correctly", () => {
    const store = mockStore(initialState);

    const { getByText } = render(
      <Provider store={store}>
        <NavBar isSignedIn={false} appName={appName} />
      </Provider>
    );

    expect(getByText(appName)).toMatchSnapshot();
  });

  it("renders UserActions", () => {
    const store = mockStore(initialState);

    const { getByText } = render(
      <Provider store={store}>
        <NavBar isSignedIn={false} appName={appName} />
      </Provider>
    );

    expect(getByText("Register")).toBeTruthy();
  });

  it("maps state properties correctly", () => {
    const userState: UserState = {
      isSignedIn: true,
      isAuthenticating: false,
    };

    const pipelinerState = { ...initialState, user: userState };
    const store = mockStore({ ...pipelinerState, isSignedIn: true });

    const { queryByTestId } = render(
      <Provider store={store}>
        <NavBarContainer appName="TestNavBar" />
      </Provider>
    );

    expect(queryByTestId("register-dialog-button")).toBeNull();
  });

  it("dispatches actions to store on register", async () => {
    UserService.prototype.registerUser = (): Promise<string> =>
      new Promise((res: Function) => res("authToken"));

    const mockStore = configureStore([thunk]);
    const store = mockStore({
      ...initialState,
      ui: { showRegisterModal: true, showSignInModal: false },
    });

    const { getByTestId } = render(
      <Provider store={store}>
        <NavBarContainer appName="TestNavBar" />
      </Provider>
    );

    fireEvent.change(getByTestId("username-input"), {
      target: { value: username },
    });

    fireEvent.change(getByTestId("emailAddress-input"), {
      target: { value: emailAddress },
    });

    fireEvent.change(getByTestId("password-input"), {
      target: { value: password },
    });

    fireEvent.click(getByTestId("register-button"));

    const actions = await store.getActions();

    await expect(actions).toEqual([
      { isAuthenticating: true, type: SET_IS_AUTHENTICATING },
      { type: SIGN_IN },
      { successMessage: "Registration Successful!", type: SHOW_SUCCESS },
      { show: false, type: SET_SHOW_REGISTER_MODAL },
      { isAuthenticating: false, type: SET_IS_AUTHENTICATING },
    ]);
  });
});
