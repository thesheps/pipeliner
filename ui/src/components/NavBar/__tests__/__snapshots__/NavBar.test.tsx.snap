// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar renders correctly 1`] = `
<h6
  class="MuiTypography-root makeStyles-title-3 MuiTypography-h6"
>
  My Lovely Test
</h6>
`;
